### PR TITLE
SNOW-1332640 support reading of structured types

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [ ] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [ ] Create tests which fail without the change (if possible)
- [ ] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [ ] Extend the README / documentation and ensure is properly displayed (if necessary)
- [ ] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/connection/result/column.js [hunk 1]
```diff
@@ -268,14 +276,89 @@ function convertRawBigInt(rawColumnValue) {
...
+  case 'boolean':
+    value =  convertRawBoolean(columnValue);
+    break;
+  case 'timestamp_ltz':
+    context.format = context.statementParameters['TIMESTAMP_LTZ_OUTPUT_FORMAT'] || context.statementParameters['TIMESTAMP_OUTPUT_FORMAT'];
+    value = convertDateTimeString(columnValue, context.format, context.statementParameters['TIMEZONE'], column.getScale()).toSfDate();
+    break;
+  case 'timestamp_ntz':
+    context.format = context.statementParameters['TIMESTAMP_NTZ_OUTPUT_FORMAT'] || context.statementParameters['TIMESTAMP_OUTPUT_FORMAT'];
+    value = convertDateTimeString(columnValue, context.format, moment.tz.guess(), column.getScale()).toSfDate();
```
Comment (sfc-gh-pfus) [inline lib/connection/result/column.js:329]: Is `guess` ok? It should be no timezone here.

File: lib/connection/result/column.js [hunk 2]
```diff
@@ -268,14 +276,89 @@ function convertRawBigInt(rawColumnValue) {
...
+    break;
+  case 'timestamp_tz':
+    context.format = context.statementParameters['TIMESTAMP_TZ_OUTPUT_FORMAT'] || context.statementParameters['TIMESTAMP_OUTPUT_FORMAT'];
+    value = convertDateTimeString(columnValue, context.format, context.statementParameters['TIMEZONE'], column.getScale()).toSfDate();
```
Comment (sfc-gh-pfus) [inline lib/connection/result/column.js:333]: Why session timezone? It should be a value from timestamp itself?

File: test/integration/testStructuredType.js [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+          executeOptions.sqlText = selectObject;
+          executeOptions.complete = function (err, stmt) {
+            assert.ok(!err, JSON.stringify(err));
+            let rowCount = 0;
+            const stream = stmt.streamRows();
+
+            stream.on('readable', function () {
+              let row;
+              while ((row = stream.read()) !== null) {
+                const narmalizedRow = {};
```
Comment (sfc-gh-pfus) [inline test/integration/testStructuredType.js:67]: Typo.

File: test/integration/testStructuredType.js [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+                const expected = {};
+                Object.keys(row).forEach((key) => {
+                  narmalizedRow[key] = testUtil.normalizeRowObject(row[key]);
+                });
+                expected.timestamp_ltz = '2021-12-22 09:43:44.000 -0800';
```
Comment (sfc-gh-pfus) [inline test/integration/testStructuredType.js:72]: Why do we have a timezone here? Is it something that is returned to the customer?

File: lib/connection/result/column.js (hunk 1 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/result/column.js:329]: Changed to UTC.

File: lib/connection/result/column.js (hunk 2 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/result/column.js:333]: The timezone is used later to create correct string representation in method toJSON

File: test/integration/testStructuredType.js (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testStructuredType.js]: fixed

File: test/integration/testStructuredType.js [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        '\'binary\': TO_BINARY(\'616263\', \'HEX\') ' +
+        '}' +
+        '::OBJECT(string VARCHAR' +
+        ', b TINYINT, ' +
+        's SMALLINT, ' +
+        'i INTEGER, ' +
+        'l BIGINT, ' +
+        'f FLOAT, ' +
+        'd DOUBLE, ' +
+        'bd DOUBLE, ' +
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply test/integration/testStructuredType.js:283]: It should be big number

File: lib/connection/result/datetime_format_converter.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    ['TZHTZM', 'ZZ'],
+
+    // special code needed
+    ['TZH', ''],
+    ['TZM', ''],
+    ['FF', '']
+  ];
+}
+
+function convertSnowflakeFormatToMomentFormat(formatSql, scale) {
```
Comment (sfc-gh-pfus) [inline lib/connection/result/datetime_format_converter.js:31]: sqlFormat?

File: lib/connection/result/sf_timestamp.js [hunk 7]
```diff
@@ -160,10 +132,19 @@ SfTimestamp.prototype.toString = function () {
...
+  if (timezone) {
+    if (typeof timezone === 'number') {
+      this._valueAsString = moment.utcOffset(timezone).format(formatMoment);
+    } else {
+      this._valueAsString = moment.tz(timezone).format(formatMoment);
+    }
+  } else {
+    this._valueAsString = moment.format(formatMoment);
+  }
+  Logger.getInstance().info(`Format: ${formatMoment}`);
```
Comment (sfc-gh-pfus) [inline lib/connection/result/sf_timestamp.js:146]: Logs to remove?

File: test/unit/connection/result/datetime_format_converter_test.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      inputFormat: 'MMMM DD, YYYY DY HH24:MI:SS.FF9 TZH:TZM',
+      output: 'MMMM DD, YYYY ddd HH:mm:ss. Z',
+    },
+    {
+      inputFormat: 'MON DD, YYYY HH12:MI:SS,FF9PM TZH:TZM',
+      output: 'MMM DD, YYYY hh:mm:ss,A Z',
+    },
+    {
+      inputFormat: 'HH24:MI:SS.FF3 HH12:MI:SS,FF9',
+      output: 'HH:mm:ss. hh:mm:ss,',
```
Comment (sfc-gh-pfus) [inline test/unit/connection/result/datetime_format_converter_test.js:28]: Why second fraction is dropped?

File: lib/connection/result/column.js [hunk 9]
```diff
@@ -268,14 +276,108 @@ function convertRawBigInt(rawColumnValue) {
...
+    value =  toValueFromNumber(convertRawNumber(columnValue));
+    break;
+  case 'boolean':
+    value =  convertRawBoolean(columnValue);
+    break;
+  case 'timestamp_ltz':
+    value = convertTimestampTzString(columnValue, formatLtz, context.statementParameters['TIMEZONE'], column.getScale()).toSfDate();
+    break;
+  case 'timestamp_ntz':
+    // context.format = context.statementParameters['TIMESTAMP_NTZ_OUTPUT_FORMAT'] || context.statementParameters['TIMESTAMP_OUTPUT_FORMAT'];
```
Comment (sfc-gh-dheyman-1) [inline lib/connection/result/column.js:332]: can we delete this comment?

File: test/integration/testStructuredType.js
Comment (sfc-gh-dheyman-1) [inline test/integration/testStructuredType.js:1]: what is the file naming convention? I see that we use snake_case and camelCase at the same time

File: test/integration/testStructuredType.js [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+              callback();
+            }
+          });
+        },
+      ],
+      done
+      );
+    });
+
+    // it('test nested object', function (done) {
```
Comment (sfc-gh-dheyman-1) [inline test/integration/testStructuredType.js:393]: why commented? do we need this?

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

File: test/integration/testConnection.js [hunk 11]
```diff
@@ -46,7 +46,7 @@ describe('Connection test', function () {
     assert.ok(!connection.isUp(), 'still active');
   });
 
-  it('Wrong Username', function (done) {
+  it.skip('Wrong Username', function (done) {
```
Comment (sfc-gh-pbulawa) [inline test/integration/testConnection.js:49]: Why do we skip this test?

File: test/integration/testStructuredType.js [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+        '::OBJECT(timestampltz TIMESTAMP_LTZ) As RESULT';
+      const expected = { timestampltz: '2021-12-22 09:43:44.000 -0800' };
+      async.series([
+        function (callback) {
+          connection.execute({
+            sqlText: selectObject,
+            complete: function (err, stmt, rows) {
+              testUtil.checkError(err);
+              const row = rows[0];
+              const narmalizedRow = {};
```
Comment (sfc-gh-pbulawa) [inline test/integration/testStructuredType.js:67]: Typo `narmalized` instead of `normalized`.

File: test/integration/testStructuredType.js [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+      const expected = { timestampntz: '2021-12-22 09:43:44.000' };
+
+      async.series([
+        function (callback) {
+          connection.execute({
+            sqlText: selectObject,
+            complete: function (err, stmt, rows) {
+              testUtil.checkError(err);
+              const row = rows[0];
+              const narmalizedRow = {};
```
Comment (sfc-gh-pbulawa) [inline test/integration/testStructuredType.js:124]: Typo

**Review by sfc-gh-pbulawa — APPROVED** [review]

**Comment by sfc-gh-pmotacki** [comment]
Added to changes in feather branch - PR #874

**Pull request closed without merging**